agent/cache/watch.go

```diff
@@ -0,0 +1,103 @@
+package cache
+
+import (
+	"context"
+	"fmt"
+	"time"
+)
+
+// WatchUpdate is a struct summarising an update to a cache entry
+type WatchUpdate struct {
+	Result interface{}
+	Meta   ResultMeta
+	Err    error
+}
+
+// indexManagedRequest is a wrapper around a user-passed Request that allows us
+// to manage the blocking index without knowing how to modify the concrete
+// Request type.
+type indexManagedRequest struct {
+	r        Request
+	minIndex uint64
+}
+
+// CacheInfo implements Request
+func (r *indexManagedRequest) CacheInfo() RequestInfo {
+	info := r.r.CacheInfo()
+	info.MinIndex = r.minIndex
+	info.Timeout = 10 * time.Minute
+	return info
+}
+
+// Watch behaves like Get but return a chan to be watched for updates
+// asynchronously. It is a helper that abstracts code from perfroming their own
+// "blocking" query logic against a cache key to watch for changes and to
+// maintain the key in cache actively. It will continue to perform blocking Get
+// requests until the context is canceled.
```

Comment:
Probably worth documenting that the watcher will block if the channel write blocks, so care should be taken by the reader to read quickly.

Reply:
Good catch.

Reply:
Added a comment (incoming) but it turns out this is probably never a problem - worst that can happen is that no cache Get's are made for long enough for the result to be evicted, but even then as soon as the caller drains the chan again, new fetching will resume...

Most important bad behaviour to document it not leaving the watch without ever cancelling it as that leaks the goroutine.